### ISSUE TITLE
MapClick Teleport improvements

### ISSUE THREAD

**Issue by shadow-fa** [Raptoer/F3#130]
~~There's slight "issue" (which is probably not even measurable) with MapClick Teleport:
`https://github.com/Raptoer/F3/blob/f21f1a696d5a61460d726adba6e7fdc5785f5f70/f/mapClickTeleport/fn_mapClickTeleportUnit.sqf#L18-L20`
Here, it is assumed that the player clicks on the map after activating the action menu `Teleport`/`HALO`. But if the player would just close the map afterwards, then the `waitUntil` would continue to check the condition at every frame, potentially forever (if the action will not be done before the timer runs out).~~
Edit: Handled in #157

### New feature: Teleporting vehicles

I think we should give missionmakers the option to teleport entire vehicles. We've had some missions with this already and it seemed to work just fine. Of course there are some caveats which we'd have to warn about (in the wiki). E.g. teleporting after safe start ended may be dangerous because of trees/rocks.

**Comment by NikkoJT** [comment]
I'll leave the waitUntil issue to those who understand such things.

Teleporting vehicles would certainly be a useful feature. I can't see any reason not to have it as an option.

**Comment by shadow-fa** [comment]
When teleporting vehicles, some things have to be considered.
* If the height is `>0` (HALO), should we still teleport the vehicle? Should be keep everyone seated and attach a parachute? If so, when should it open? However, it should be noted that, unlike parachutes that are attached to units, the vehicle-parachute is not controllable.
    Several other options have been brought up in #157:
    1. As now (the group leader is in a vehicle, only the men are HALO'ed and not the vehicle.) but the (empty) vehicle is also teleported to the position, with a similar dispersal, and a parachute attached immediately. Issue being it'll take a long time to land.
    2. Teleport vehicles with men inside and an automatically opening chute. Issue being complexity of auto-opening chute (is there a vanilla object / function for this?)
    3. Teleport vehicles with men inside and an action for the teleport starter to open the chute. Maybe best but could lead to 'pancaked' vehicles.
    4. ignore the height and teleport them on the ground.
    5. keep it like it currently is, because it's always been like this. Changing this behaviour could break missions?
* If `f_var_mapClickTeleport_GroupTeleport = false`, any passengers will still be able to teleport themselves (and the vehicle!) after the vehicle has been teleported by someone in the vehicle. The same is true if `f_var_mapClickTeleport_Uses` is not exactly `1`.

Currently the component has the following variables in `init.sqf`
```sqf
// f_var_mapClickTeleport_Uses = 0;              // How often the teleport action can be used. 0 = infinite usage.
// f_var_mapClickTeleport_TimeLimit = 0;         // If higher than 0 the action will be removed after the given time.
// f_var_mapClickTeleport_GroupTeleport = false; // False: everyone can teleport. True: Only group leaders can teleport and will move their entire group.
// f_var_mapClickTeleport_Units = [];            // Restrict map click teleport to these units
// f_var_mapClickTeleport_Height = 0;            // If > 0 map click teleport will act as a HALO drop and automatically assign parachutes to units
```
@SamLex suggested to think about removing some of these options (To make the code simpler). E.g. `f_var_mapClickTeleport_GroupTeleport` is rarely set to `false`, so it could be removed entirely.
However, changing existing parameters might not be so good, because existing missions might break when updating to a newer version of the framework.

* Not really related to vehicles, but I was thinking about adding a new feature which would allow teleportation if the unit/group is inside a given area. (This could also maybe replace the `Uses` and `TimeLimit` parameter).

* Regarding removing options: I will go through all missions we have on famdb, to see if any missions actually use the `f_var_mapClickTeleport_Uses`, `f_var_mapClickTeleport_GroupTeleport`, and `f_var_mapClickTeleport_Units` variables. Once I have the data, I will update this post (or reply to this thread)

**Comment by shadow-fa** [comment]
I have analyzed all existing FA missions.
Most of them use the following settings:
```
GroupTeleport = true
Uses = 1
Units = []
```
These will become the new default values, so that missionmakers don't have to change them for every single mission.


The only missions that have a different value for `Uses` are workshop missions.
They use `Uses = 0` (= infinity).
This applies to the following missions:
```
fa3_w36_vehicle_command_v2.Altis
fa3_w40_Workshop_v4.Altis
fa3_workshop_strict_v2.Altis
```

Missions that actually use the `Units` variable:
```
fa3_a128_cratesistance_v3.Altis:     f_var_mapClickTeleport_Units = ["UnitFIA_CO","UnitFIA_ASL_SL","UnitFIA_BSL_SL","UnitFIA_CSL_SL","UnitFIA_A2_FTL","UnitFIA_B2_FTL","UnitFIA_C2_FTL"];
fa3_a28_alternative_financing_v1.Altis:f_var_mapClickTeleport_Units = ["UnitFIA_ASL_SL", "UnitFIA_ASL_M", "UnitFIA_A1_FTL", "UnitFIA_A1_AR1", "UnitFIA_A1_AR2", "UnitFIA_A1_AT", "UnitFIA_A1_R1", "UnitFIA_A1_R2", "UnitFIA_A2_FTL", "UnitFIA_A2_AR1", "UnitFIA_A2_AR2", "UnitFIA_A2_AT", "UnitFIA_A2_R1", "UnitFIA_A2_R2"];
fa3_a55_roadhouse_rendezvous_v1.Tanoa:{ if (side _x == independent) then {f_var_mapClickTeleport_Units pushBack _x}} forEach playableUnits;
fa3_a78_missed_connections_v1.Tanoa:f_var_mapClickTeleport_Units = _synUnits;
fa3_a82_wetwork_v1.Tanoa:            f_var_mapClickTeleport_Units pushBack _x;
fa3_a88_cratesistance_tanoa_v2.Tanoa:        f_var_mapClickTeleport_Units = ["UnitSyn_CO","UnitSyn_ASL_SL","UnitSyn_BSL_SL","UnitSyn_CSL_SL","UnitSyn_A2_FTL","UnitSyn_B2_FTL","UnitSyn_C2_FTL","UnitSyn_IFV1_D"];
fa3_c42_city_watch_v1.Altis:f_var_mapClickTeleport_Units = ["UnitNATO_CO","UnitNATO_ASL_SL","UnitNATO_A1_FTL","UnitNATO_A2_FTL","UnitNATO_IFV1_C","UnitAAF_CO","UnitAAF_ASL_SL","UnitAAF_A1_FTL","UnitAAF_A2_FTL","UnitAAF_IFV1_C"];
fa3_c76_citadel_v9.Altis: f_var_mapClickTeleport_Units = ["UnitCSAT_ASL_SL","UnitCSAT_A1_FTL","UnitCSAT_A2_FTL"];
```
However, most (if not all) of them only use this variable to 
limit the teleport option to a single faction.
This can also be done in a different way.

The last variable `GroupTeleport` is only set to `false` for the following missions:
```
fa3_a128_cratesistance_v3.Altis
fa3_a88_cratesistance_tanoa_v2.Tanoa
fa3_w40_Workshop_v4.Altis
fa3_workshop_strict_v2.Altis
```

While this means that most options aren't really needed,
I don't think we should remove them entirely
as it would become more difficult to update older missions.

The parameters `GroupTeleport` and `Uses` are necessary for workshop missions.
`Units` might not be needed, but it could be useful for future features like vehicle teleportations (e.g. if a group is split up into multiple vehicles)

**Comment by shadow-fa** [comment]
Details about handling vehicles:
- When inside a vehicle that itself is inside another vehicle (vehicle-in-vehicle transport), the action will do nothing.
- When group members are not in the same vehicle, they are teleported alongside the vehicle and have their own parachute

Other implementation details as currently implemented in #157:
- A new briefing page will be shown when HALO jumping is enabled.
